src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -769,13 +686,15 @@ public Solution RemoveMetadataReference(ProjectId projectId, MetadataReference m
                 throw new ArgumentNullException(nameof(metadataReference));
             }
 
-            var newState = _state.RemoveMetadataReference(projectId, metadataReference);
-            if (newState == _state)
+            // If the project didn't change itself, there's no need to change the compilation state.
+            var stateChange = _state.RemoveMetadataReference(projectId, metadataReference);
+            if (stateChange.NewSolutionState == _state)
```

Comment:
This did jump out to me when looking at SolutionCompilationState as to why RemoveMetadataReferences took the state change unlike everything else -- it might be a bit less confusing to just have the Remove method here directly validate the reference exists, and then make everything else consistent.

Reply:
done in followup.